src/axolotl/state.rs

```diff
@@ -144,22 +143,25 @@ impl <T:Axolotl> ReceiveChain<T> {
         let (_,ref message_key) = self.message_keys[message_key_index];
         let expected_mac = axolotl_impl.authenticate_message(message, message_key, sender_identity, receiver_identity);
         if expected_mac == mac {
-            axolotl_impl.decrypt_message(&message_key, &message.ciphertext)
+            axolotl_impl.with_decoded_ciphertext(message, |ciphertext| {
```

Comment:
the problem here is that the decoded ciphertext could just be a part of the message, or it could be something on the heap, or it could be a local variable of decode.

if traits allowed generic associated types (like they allow generic functions), there would be a much more elegant solution: Ciphertext could have a lifetime parameter and implement the Borrow trait. it's a planned/desired feature, but it's not there yet.

there might be some type system trickery to work around it, but a CPS transformation is the best i have come up with so far.